File: rede-bayseana-smile/tutorials/R/tutorial6.R
library(rSMILE)
source("License.R")

# Tutorial6 creates a dynamic Bayesian network (DBN),
# performs the inference, then saves the model to disk.

createCptNode = function(net, id, name, outcomes, xPos, yPos) {
    handle <- net$addNode(net$NodeType$CPT, id)
    initialOutcomeCount <- net$getOutcomeCount(handle)
    count <- length(outcomes)
    net$setNodeName(handle, name)
    net$setNodePosition(handle, xPos, yPos, 85L, 55L)
    if (!is.null(outcomes)) {
        sapply(0:(initialOutcomeCount-1), 
            function(x) net$setOutcomeId(handle, x, outcomes[x+1]))
        if (initialOutcomeCount < count) {
            sapply(initialOutcomeCount:(count-1), 
                function(x) net$addOutcome(handle, outcomes[x+1]))
        }
    }

    return(handle)
}

updateAndShowTemporalResults = function(net) {
    net$updateBeliefs()
    sliceCount <- net$getSliceCount()
    handles <- net$getAllNodes()
    for (h in handles) {
        if (net$getNodeTemporalType(h) == net$NodeTemporalType$PLATE) {
            outcomeCount <- net$getOutcomeCount(h)
            cat(sprintf("Temporal beliefs for %s:\n", net$getNodeId(h)))
            v <- net$getNodeValue(h)
            for (sliceIdx in 0:(sliceCount-1)) {
                cat(sprintf("\tt=%d:", sliceIdx))
                for (i in 0:(outcomeCount-1)) {
                    cat(sprintf(" %f", v[(sliceIdx*outcomeCount+i)+1]))
                }
                cat("\n")
            }
        }
    }
    cat("\n")
}

cat("Starting Tutorial6...\n")
net <- Network()

loc <- createCptNode(net, "Location", "Location", c("Pittsburgh", "Sahara"), 160, 360)
rain <- createCptNode(net, "Rain", "Rain",c("true", "false"), 380, 240)
umb <- createCptNode(net, "Umbrella", "Umbrella", c("true", "false"), 300, 100)

net$setNodeTemporalType(rain, net$NodeTemporalType$PLATE)
net$setNodeTemporalType(umb, net$NodeTemporalType$PLATE)

net$addArc(loc, rain)
net$addTemporalArc(rain, rain, 1)
net$addArc(rain, umb)

rainDef <- c( 0.7,
              0.3,
              0.01,
              0.99 )
              
net$setNodeDefinition(rain, rainDef)
              
rainDefTemporal <- c( 0.7,
                      0.3,
                      0.3,
                      0.7,
                      0.001,
                      0.999,
                      0.01,
                      0.99 )
                      
net$setNodeTemporalDefinition(rain, 1, rainDefTemporal)

umbDef <- c( 0.9,
             0.1,
             0.2,
             0.8 )
             
net$setNodeDefinition(umb, umbDef)

net$setSliceCount(5)

cat("Performing update without evidence.\n")
updateAndShowTemporalResults(net)

cat("Setting Umbrella[t=1] to true and Umbrella[t=3] to false.\n")
net$setTemporalEvidence(umb, 1, 0)
net$setTemporalEvidence(umb, 3, 1)
updateAndShowTemporalResults(net)

net$writeFile("tutorial6.xdsl")
cat("Tutorial6 complete: Network written to tutorial6.xdsl\n")

File: rede-bayseana-smile/tutorials/cpp/tutorial6.cpp
// tutorial6.cpp
// Tutorial6 creates a dynamic Bayesian network (DBN),
// performs the inference, then saves the model to disk.

#include "/home/josegustavo/Documentos/IC/rede-bayseana/include/smile.h"
#include "/home/josegustavo/Documentos/IC/rede-bayseana/include/smile_license.h"

#include <cstdio>


static int CreateCptNode(
    DSL_network& net, const char* id, const char* name,
    std::initializer_list<const char*> outcomes, int xPos, int yPos);

static void UpdateAndShowTemporalResults(DSL_network &net);


int Tutorial6()
{
    printf("Starting Tutorial6...\n");

    DSL_errorH().RedirectToFile(stdout);
    
    DSL_network net;

    int loc = CreateCptNode(net, "Location", "Location",
        { "Pittsburgh", "Sahara" }, 160, 360);

    int rain = CreateCptNode(net, "Rain", "Rain",
        { "true", "false" }, 380, 240);

    int umb = CreateCptNode(net, "Umbrella", "Umbrella",
        { "true", "false" }, 300, 100);

    net.SetTemporalType(rain, dsl_temporalType::dsl_plateNode);
    net.SetTemporalType(umb, dsl_temporalType::dsl_plateNode);

    net.AddArc(loc, rain);
    net.AddTemporalArc(rain, rain, 1);
    net.AddArc(rain, umb);

    DSL_nodeDef *rainDef = net.GetNode(rain)->Def();
    int res = rainDef->SetDefinition({
        0.7,  // P(Rain=true |Location=Pittsburgh)
        0.3,  // P(Rain=false|Location=Pittsburgh)
        0.01, // P(Rain=true |Location=Sahara)
        0.99  // P(Rain=false|Location=Sahara)
    });
    if (DSL_OKAY != res)
    {
        return res;
    }

    res = rainDef->SetTemporalDefinition(1, {
        0.7,   // P(Rain=true |Location=Pittsburgh,Rain[t-1]=true)
        0.3,   // P(Rain=false|Location=Pittsburgh,Rain[t-1]=true)
        0.3,   // P(Rain=true |Location=Pittsburgh,Rain[t-1]=false)
        0.7,   // P(Rain=false|Location=Pittsburgh,Rain[t-1]=false)
        0.001, // P(Rain=true |Location=Sahara,Rain[t-1]=true)
        0.999, // P(Rain=false|Location=Sahara,Rain[t-1]=true)
        0.01,  // P(Rain=true |Location=Sahara,Rain[t-1]=false)
        0.99   // P(Rain=false|Location=Sahara,Rain[t-1]=false)
    });
    if (DSL_OKAY != res)
    {
        return res;
    }

    res = net.GetNode(umb)->Def()->SetDefinition({
        0.9, // P(Umbrella=true |Rain=true)
        0.1, // P(Umbrella=false|Rain=true)
        0.2, // P(Umbrella=true |Rain=false)
        0.8  // P(Umbrella=false|Rain=false)
    });
    if (DSL_OKAY != res)
    {
        return res;
    }

    net.SetNumberOfSlices(5);

    printf("Performing update without evidence.\n");
    UpdateAndShowTemporalResults(net);

    printf("Setting Umbrella[t=1] to true and Umbrella[t=3] to false.\n");
    res = net.GetNode(umb)->Val()->SetTemporalEvidence(1, 0);
    if (DSL_OKAY != res)
    {
        return res;
    }
    res = net.GetNode(umb)->Val()->SetTemporalEvidence(3, 1);
    if (DSL_OKAY != res)
    {
        return res;
    }
    UpdateAndShowTemporalResults(net);

    res = net.WriteFile("tutorial6.xdsl");
    if (DSL_OKAY != res)
    {
        return res;
    }
    
    printf("Tutorial6 complete: Network written to tutorial6.xdsl\n");
    return DSL_OKAY;
}


static void UpdateAndShowTemporalResults(DSL_network &net)
{
    net.UpdateBeliefs();
    int sliceCount = net.GetNumberOfSlices();
    for (int h = net.GetFirstNode(); h >= 0; h = net.GetNextNode(h))
    {
        if (net.GetTemporalType(h) == dsl_temporalType::dsl_plateNode)
        {
            DSL_node *node = net.GetNode(h);
            int outcomeCount = node->Def()->GetNumberOfOutcomes();
            printf("Temporal beliefs for %s:\n", node->GetId());
            const DSL_Dmatrix& mtx = *node->Val()->GetMatrix();
            for (int sliceIdx = 0; sliceIdx < sliceCount; sliceIdx++)
            {
                printf("\tt=%d:", sliceIdx);
                for (int i = 0; i < outcomeCount; i++)
                {
                    printf(" %f", mtx[sliceIdx * outcomeCount + i]);
                }
                printf("\n");
            }
        }
    }
    printf("\n");
}


static int CreateCptNode(
    DSL_network& net, const char* id, const char* name,
    std::initializer_list<const char*> outcomes, int xPos, int yPos)
{
    int handle = net.AddNode(DSL_CPT, id);
    DSL_node* node = net.GetNode(handle);
    node->SetName(name);
    node->Def()->SetNumberOfOutcomes(outcomes);
    DSL_rectangle& position = node->Info().Screen().position;
    position.center_X = xPos;
    position.center_Y = yPos;
    position.width = 85;
    position.height = 55;
    return handle;
}
